### PR TITLE
WIP: Specialized Html DSL and Rendering for Virtual-dom

### PR DESCRIPTION
**Pull request by Zaid-Ajaj**
This PR proposes a different way to tackle virtual-dom. Namely by mapping the tree directly to `VNode<'T>` instead of `DomNode<'T> -> VNode<'T>` and by-passing `createTree` all together.  

This will live inside `Fable.Arch.Virtualdom.Html` and you would open this module to access the dsl instead of `Fable.Arch.Html` 

It is intented to fully support virtual-dom feature called [Widgets](https://github.com/Matt-Esch/virtual-dom/blob/master/docs/widget.md). This seems to be a sub-tree that virtual-dom will skip during diffing and patching allowing us to EMBED small (component-like) apps inside bigger apps without touching their dispatch loop or nesting app models. Integrations with other libraries should also be fairly easy when this is implemented.  

Also, intellisense for the attributes and the properties.

an example usage of code looks like:
```fsharp
let view : VNode<Message> = 
      div 
        [
           Attrs [
              Style "color:white"
              ClassName "content"
              Id "unique"
           ]
           OnClick (fun ev -> Decr)
        ] 
        [
          div [] [ text "First div"  ]
          div [] [ text "Second div" ]
          div [] [ text "Third div"  ]
        ]
```
Feedbacks are welcome 😄

### CODE REVIEW

File: src/Fable.Arch/Fable.Arch.Virtualdom.Html.fs [hunk 1]
```diff
@@ -0,0 +1,369 @@
...
+
+
+
+type VNode<'T> = 
+    | Plain of string
+    | Unary of VTag * VProps<'T> list
+    | Nested of VTag * VProps<'T> list * VNode<'T> list
+
+
+/// Example usage
```
Comment (MangelMaxime) [inline src/Fable.Arch/Fable.Arch.Virtualdom.Html.fs:348]: Not sure that the example usage should go here.

I would prefer an independant sample showing how to use this feature.

**Comment by MangelMaxime** [comment]
With this implementation we will have a different DSL for the "Widgets' views" and for the classic app views. This can be confusion for the user no ?

File: src/Fable.Arch/Fable.Arch.Virtualdom.Html.fs (hunk 1 above)
Comment (Zaid-Ajaj) [reply src/Fable.Arch/Fable.Arch.Virtualdom.Html.fs:348]: Agree, it was just there to prove the DSL compiles 😁

**Comment by Zaid-Ajaj** [comment]
The idea is to use the same DSL for the view and the widgets. In Virtual dom widgets, we deal with `HTMLElement`s directly which we can wrap to let the user deal with just `VNode<'T>`. The user would then create a widget just the way you create an app but you can embed it inside other bigger apps. 

React uses `ReactElement`s, why not let virtual-dom do it's own thing too with `VNode<T>`

**Comment by MangelMaxime** [comment]
What I mean is:

This code :
```fsharp
      div 
        [
           Attrs [
              Style "color:white"
              ClassName "content"
              Id "unique"
           ]
           OnClick (fun ev -> Decr)
        ]
```

would have been written like this (in Fable-arch):

```fsharp
      div 
        [
           Style [ "color", "white"]
           property "className" "content"
           property "id" "unique"
           onMouseClick (fun _ -> Decr)
        ]
```

File: src/Fable.Arch/Fable.Arch.Virtualdom.Html.fs (hunk 1 above)
Comment (MangelMaxime) [reply src/Fable.Arch/Fable.Arch.Virtualdom.Html.fs:348]: Ok no problem.  :D
If you could add an independant sample that would be great then. Just to help see how everything is fitting :)

**Comment by Zaid-Ajaj** [comment]
Yes but then you are getting `DomNode<'T>` that you would have to convert to `VNode<'T>` with `createTree`. using the above DSL, there would be no need for `createTree` and the tree is directly usable for `diff` and `patch` operations. Also, you don't have to remove the current Html dsl. They Live in different modules.

**Comment by MangelMaxime** [comment]
I have something bothering me.

We are tending in a near future to "unify" Fable-Arch and Fable-Elmish. IMO adding this kind of feature into Fable-Arch is somehow going the opposite way.

Yes this can be a cool feature and available as it is only in VirtualDom but the new renderer Snabbdom for Fable-Elmish give access to more hooks and should allow to achieve the same goal at the end according to this issue [https://github.com/snabbdom/snabbdom/issues/28](https://github.com/snabbdom/snabbdom/issues/28).

After, we also need to think if this is really something we want to make available. I am not sure of the benefits and perhaps this could tend to bad/dangerous practice by users.

Personally I would prefer to limit the new feature in Fable-Arch and focus on making Elmish-Snabbdom renderer alpha ready. (At the moment, I got one last blocking bug).

In my opinion, if we start adding this kind of feature to Fable-Arch we will have to keep maintaining both project and keep taking inspiration from one to the other.

Could we compare our opinion on the subject ? @alfonsogarciacaro @mastoj @et1975 

PS: @Zaid-Ajaj I am not saying no ^^ just trying to figure/discuss what is the best way to go.

**Comment by MangelMaxime** [comment]
Just adding a message I share with Eugene:

> Just wanting to share a thought I was having when working on Fable-Arch and Elmish merge.
> The big difference I see now between both of them is:
> 
>     1. Fable-Arch is plugin centric
>     2. Elmish is app centric.
> 
> What I mean is when we need the navigation module into Fable-Arch we are adding it more like a plugin thanks to the Producer/ Subscriber.
> Where in Elmish this just an NavigationApp wrapping the user application.
> Not sure of the benefit of this though but I think it was worth telling to help deciding merging both project or not when ready ^^.

And perhaps Elmish is the framework to use when you want to follow a more true Elm architecture and Fable-Arch an "hybrid" framework.

**Comment by mastoj** [comment]
@MangelMaxime I think I agree with you. I think that the Fable community will be better of if we manage to unify Arch with elmish somehow. 

Also, you description sounds accurate. My intention when I began writing Arch was to be able to plugin in various places in the application, and you desciption of elmish is my view of it too. I don't think those approaches are mutually exclusive though and could probably be combined someway.

**Comment by alfonsogarciacaro** [comment]
TBH, I haven't played with fable-elmish and fable-arch as much as I would have liked to, so you know much more than me about the strengths of each other. But if we are to unify them I'd recommend to do it sooner rather than later. This is because I want to release Fable 1.0 soon and start focusing on promotion, samples, tutorials, etc. It'd be great if for that time we have one single architecture that we can recommend users and include in templates. I'm pretty sure that you'll also get many good contributions from the community (as the projects are already getting independently).

So, in conclusion. I wouldn't think too much of the difficulties to merge both projects, I'd just do it and later we can apply solutions to any problem that may appear.

I'd try to help as much as possible. @MangelMaxime, what's the blocking bug you mentioned?

**Comment by Zaid-Ajaj** [comment]
@MangelMaxime This PR bothering you? BAD PR! closing.... 😄 

I agree with this all. Moving now (mentally) to fable-elmish
